### ISSUE TITLE
Crash when registering 2 custom physics servers of same type(eg. 2d) or more at a time

### ISSUE THREAD

**Issue by Ughuuu** [godotengine/godot#83828]
### Godot version

4.2

### System information

macOS m1 air

### Issue description

Register box2d and rapier2d. Crash in rapier2d(As it's second one registered).

### Steps to reproduce

Install from asset store box2d and rapier2d. Restart project, when starting results in crash.

### Minimal reproduction project

Follow repro steps as the size of binaries could add up to a rather big zip file.

**Comment by Ughuuu** [comment]
FYI the use case is comparing the two physics engines without having to delete every time the other one from addons folder.

**Comment by akien-mga** [comment]
> Godot version
>
> 4.2

Please be more specific, 4.2 spans 4 months of development. I assume you mean "latest" but latest is a fast moving target.

I tested on Linux with 8c25a98fdf68dc388eddaefa188f3386236cdfe4 and it's not crashing for me when I follow the steps to reproduce.

I do get a number of errors when opening the project in the editor, but no crash:
```
ERROR: Parameter "body" is null.
   at: _get_step (src/bodies/box2d_direct_body_state.cpp:173)
ERROR: Parameter "body" is null.
   at: _get_inverse_mass (src/bodies/box2d_direct_body_state.cpp:29)
ERROR: Parameter "body" is null.
   at: _get_inverse_inertia (src/bodies/box2d_direct_body_state.cpp:33)
ERROR: Parameter "body" is null.
   at: _get_total_angular_damp (src/bodies/box2d_direct_body_state.cpp:14)
ERROR: Parameter "body" is null.
   at: _get_total_linear_damp (src/bodies/box2d_direct_body_state.cpp:9)
ERROR: Parameter "body" is null.
   at: _get_total_gravity (src/bodies/box2d_direct_body_state.cpp:4)
ERROR: Parameter "body" is null.
   at: _get_center_of_mass (src/bodies/box2d_direct_body_state.cpp:19)
ERROR: Parameter "body" is null.
   at: _get_center_of_mass_local (src/bodies/box2d_direct_body_state.cpp:24)
ERROR: Parameter "body" is null.
   at: _get_angular_velocity (src/bodies/box2d_direct_body_state.cpp:49)
ERROR: Parameter "body" is null.
   at: _get_linear_velocity (src/bodies/box2d_direct_body_state.cpp:41)
ERROR: Parameter "body" is null.
   at: _is_sleeping (src/bodies/box2d_direct_body_state.cpp:121)
ERROR: Parameter "body" is null.
   at: _get_transform (src/bodies/box2d_direct_body_state.cpp:57)
ERROR: Unsupported shape
   at: _shape_create (src/rapier_physics_server_2d.cpp:40)
ERROR: Condition "!shape" is true.
   at: _shape_set_data (src/rapier_physics_server_2d.cpp:83)
ERROR: Invalid ID.
   at: _free_rid (src/rapier_physics_server_2d.cpp:1183)
```

Those might come from the documentation generation when it tries to instantiate the Box2D and Rapier classes to retrieve their default property values.

I tried switching to both Box2D and Rapier2D as the physics engine, and a basic free falling RigidBody2D seems to work fine.

If you can reproduce the crash in the latest `master` commit, could you attach a stacktrace? Make sure to build with debug symbols (`debug_symbols=yes`, or it's implicit if you use `dev_build=yes`).

**Comment by Ughuuu** [comment]
Sorry, I wanted to write 4.2 beta 2. It happens with latest too, eg. this commit on master: https://github.com/godotengine/godot/commit/c21c2706ad33c59295f6a3ac1e128db6fa7cce69
The crash is a:
```
Exception has occurred.
EXC_BAD_ACCESS (code=1, address=0x1d0)
```

That happens at:
```
godot.macos.editor.dev.arm64!CowData<PhysicsServer2DManager::ClassInfo>::_get_size() const (/Users/dragosdaian/Documents/godot/core/templates/cowdata.h:81)
godot.macos.editor.dev.arm64!CowData<PhysicsServer2DManager::ClassInfo>::size() const (/Users/dragosdaian/Documents/godot/core/templates/cowdata.h:134)
godot.macos.editor.dev.arm64!Vector<PhysicsServer2DManager::ClassInfo>::size() const (/Users/dragosdaian/Documents/godot/core/templates/vector.h:93)
godot.macos.editor.dev.arm64!PhysicsServer2DManager::find_server_id(String const&) (/Users/dragosdaian/Documents/godot/servers/physics_server_2d.cpp:955)
godot.macos.editor.dev.arm64!PhysicsServer2DManager::register_server(String const&, Callable const&) (/Users/dragosdaian/Documents/godot/servers/physics_server_2d.cpp:940)
godot.macos.editor.dev.arm64!void call_with_ptr_args_helper<__UnexistingClass, String const&, Callable const&, 0ul, 1ul>(__UnexistingClass*, void (__UnexistingClass::*)(String const&, Callable const&), void const**, IndexSequence<0ul, 1ul>) (/Users/dragosdaian/Documents/godot/core/variant/binder_common.h:324)
godot.macos.editor.dev.arm64!void call_with_ptr_args<__UnexistingClass, String const&, Callable const&>(__UnexistingClass*, void (__UnexistingClass::*)(String const&, Callable const&), void const**) (/Users/dragosdaian/Documents/godot/core/variant/binder_common.h:572)
godot.macos.editor.dev.arm64!MethodBindT<String const&, Callable const&>::ptrcall(Object*, void const**, void*) const (/Users/dragosdaian/Documents/godot/core/object/method_bind.h:352)
godot.macos.editor.dev.arm64!gdextension_object_method_bind_ptrcall(void const*, void*, void const* const*, void*) (/Users/dragosdaian/Documents/godot/core/extension/gdextension_interface.cpp:1133)
libphysics_server_rapier2d.macos.template_debug!godot::PhysicsServer2DManager::register_server(godot::String const&, godot::Callable const&) (Unknown Source:0)
libphysics_server_rapier2d.macos.template_debug!initialize_rapier_2d_module(godot::ModuleInitializationLevel) (Unknown Source:0)
libphysics_server_rapier2d.macos.template_debug!godot::GDExtensionBinding::initialize_level(void*, GDExtensionInitializationLevel) (Unknown Source:0)
godot.macos.editor.dev.arm64!GDExtension::initialize_library(GDExtension::InitializationLevel) (/Users/dragosdaian/Documents/godot/core/extension/gdextension.cpp:782)
godot.macos.editor.dev.arm64!GDExtensionManager::initialize_extensions(GDExtension::InitializationLevel) (/Users/dragosdaian/Documents/godot/core/extension/gdextension_manager.cpp:185)
godot.macos.editor.dev.arm64!Main::setup2() (/Users/dragosdaian/Documents/godot/main/main.cpp:2267)
godot.macos.editor.dev.arm64!Main::setup(char const*, int, char**, bool) (/Users/dragosdaian/Documents/godot/main/main.cpp:2161)
godot.macos.editor.dev.arm64!main (/Users/dragosdaian/Documents/godot/platform/macos/godot_main_macos.mm:68)
start (Unknown Source:0)
```

Looking at the stack, the first, second and third time it call `register_server` which it calls `find_server_id` here:
```
void PhysicsServer2DManager::register_server(const String &p_name, const Callable &p_create_callback) {
	//ERR_FAIL_COND(!p_create_callback.is_valid());
	ERR_FAIL_COND(find_server_id(p_name) != -1);
	physics_2d_servers.push_back(ClassInfo(p_name, p_create_callback));
	on_servers_changed();
}
```

The third time, when it adds rapier2d, it crashes(rapier2d is last as it's with letter r I think.

The crash is at line `ERR_FAIL_COND(find_server_id(p_name) != -1);`

**Comment by akien-mga** [comment]
It seems to crash in `find_server_id` when calling `physics_2d_servers.size()`, which is weird. I don't see why `Vector::size()` would crash. The line it actually crashes on is:
```cpp
	_FORCE_INLINE_ uint32_t *_get_size() const {
		if (!_ptr) {  // Crash here.
			return nullptr;
		}

		return reinterpret_cast<uint32_t *>(_ptr) - 1;
	}
```

So `CowData::_ptr` seems to hold invalid memory.

**Comment by Ughuuu** [comment]
Indeed it is, if I only have 1 extra physics server, it works. But the moment I add 2 extra ones, this happens all the time. Could be also OS specific? So far I only tasted on macOS.

**Comment by dsnopek** [comment]
The only thing I can think of is that the pointer that's being used for the `PhysicsServer2DManager` singleton on the subsequent calls isn't really a valid `PhysicsServer2DManager`, it's just some random address in memory. And if that's the case, the question is: how can one GDExtension interfere with another fetching the right pointer for a singleton? We do some caching of instances, which is based on the `godot::internal::token` in godot-cpp. Could symbols from one GDExtension be interfering with another, such that one GDExtension is using the token from another?

This line of thought reminds me of issue https://github.com/godotengine/godot/issues/83212

I wonder if @akien-mga is not having this issue on Linux is due to PR https://github.com/godotengine/godot/pull/82973

If this class of problem turns out to not just be a Linux thing (like we thought on https://github.com/godotengine/godot/issues/83212), then maybe the right solution is compiling everything (Godot itself and each GDExtension) with `-fvisibility=hidden` so that none of the symbols are named and the dynamic linker doesn't try to mess with them? (Assuming that works, I never actually tested that, I just theorized about it on the other issue.)

Anyway, sorry, that's just a pile of random guesses, more investigation needed to see if there's any substance there :-)

**Comment by Ughuuu** [comment]
Can anyone not on linux reproduce this issue though? Also, you are saying I should try building the extensions and Godot-cpp and godot with fvisibility=hidden and see if it still occurs?

**Comment by dsnopek** [comment]
> Can anyone not on linux reproduce this issue though?

I was going to try to reproduce it later when I had some more time. And if I can't reproduce it, I was going to try reverting PR https://github.com/godotengine/godot/pull/82973 locally, and see if it was reproducible then.

> Also, you are saying I should try building the extensions and Godot-cpp and godot with fvisibility=hidden and see if it still occurs?

If you have the time to do that, that'd be amazing! Otherwise, I was planning to experiment with `-fvisibility=hidden` at some point

**Comment by dsnopek** [comment]
I managed to sneak in a little bit of testing:

- On Linux, at commit f71f4b80e32f6e98a4cd3cb1c06071223297e8fc I'm also unable to reproduce the crash, everything seems to work fine. I tried reverting PR https://github.com/godotengine/godot/pull/82973 locally, but it still didn't crash.
- On Windows, I tried with Godot 4.2 beta2 - it also seemed to work fine!

In both cases, I made a little test project similar to what @akien-mga described: I put a `RigidBody2D` in the middle of the screen, with a rectangle shape and visible collision shapes turned on. I tried switching between DEFAULT, Rapier2D and Box2D, and in every case, the rigid body fell like one would expect. And no crashes, either in editor or when running the test project.

So, either:

1. This is a MacOS only problem, or
2. It's something specific to your environment somehow?

(I still plan to experiment with `-fvisibility=hidden` at some point, to see if it's a valid alternative to `RTLD_DEEPBIND`, but since I couldn't reproduce the issue on Linux even without `RTLD_DEEPBIND`, I'm not sure if this problem is even related to that.)
